python/nav/ipdevpoll/shadows/__init__.py

```diff
@@ -260,6 +267,56 @@ def _fix_binary_garbage(self):
                 setattr(self, attr, repr(value))
         self.clear_cached_objects()
 
+    @classmethod
+    def prepare_for_save(cls, containers):
+        cls.changed_devices = cls._detect_version_changes(containers)
```

Comment:
I don't approve of storing this as a class variable on `Device`. This will stick to the *class*, and not just in the data structures associated with the current job. 

This will in effect cause `Device.changed_devices` to modified concurrently by multiple collection jobs in the same process and **will most certainly have unintended side-effects and bugs in production**.

All state information regarding the current job needs to be tucked away inside `containers` - and in this case, the way to do that is to keep the information stored in the `Device` instances themselves. Adding/updating a `changed_versions` attribute on each instance would be the way to go.